// enzyme-matchers/lib/utils/html.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = printHTMLForWrapper;

var _instance = require('./instance');

var _instance2 = _interopRequireDefault(_instance);

var _getConsoleObject = require('./getConsoleObject');

var _getConsoleObject2 = _interopRequireDefault(_getConsoleObject);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var consoleObject = (0, _getConsoleObject2.default)();
var noop = function noop() {};
var error = consoleObject.error;
var SHALLOW_WRAPPER_CONSTRUCTOR = 'ShallowWrapper';

function isShallowWrapper(wrapper) {
  var isShallow = void 0;
  if (wrapper.constructor.name !== undefined) {
    isShallow = wrapper.constructor.name === SHALLOW_WRAPPER_CONSTRUCTOR;
  } else {
    // function.name isn't available on IE 11, so fall back to turning the function into
    // a string and checking its name this way.
    isShallow = !!('' + wrapper.constructor).match(/^function ShallowWrapper\(/);
  }
  return isShallow;
}

function mapWrappersHTML(wrapper) {
  return wrapper.nodes.map(function (node) {
    var inst = (0, _instance2.default)(node);
    var type = node.type || inst._tag;

    consoleObject.error = noop;

    var _ref = node.props ? node.props : inst._currentElement.props,
        children = _ref.children,
        props = _objectWithoutProperties(_ref, ['children']);

    consoleObject.error = error;

    var transformedProps = Object.keys(props).map(function (key) {
      return key + '="' + props[key] + '"';
    });
    var stringifiedNode = '<' + type + ' ' + transformedProps.join(' ');

    if (children) {
      stringifiedNode += '>[..children..]</' + node.type;
    } else {
      stringifiedNode += '/>';
    }

    return stringifiedNode;
  });
}

function printHTMLForWrapper(wrapper) {
  switch (wrapper.nodes.length) {
    case 0:
      {
        return '[empty set]';
      }
    case 1:
      {
        if (isShallowWrapper(wrapper)) {
          return wrapper.debug().replace(/\n/g, '');
        }

        return wrapper.html();
      }
    default:
      {
        var nodes = mapWrappersHTML(wrapper).reduce(function (acc, curr, index) {
          return '' + acc + index + ': ' + curr + '\n';
        }, '');

        return 'Multiple nodes found:\n' + nodes;
      }
  }
}
module.exports = exports['default'];
